File: android/src/main/kotlin/tech/shmy/a_player/player/impl/AliyunPlayerImpl.kt
package tech.shmy.a_player.player.impl

import android.content.Context
import android.view.Surface
import com.aliyun.player.AliPlayer
import com.aliyun.player.AliPlayerFactory
import com.aliyun.player.IPlayer
import com.aliyun.player.bean.InfoCode
import com.aliyun.player.source.UrlSource
import tech.shmy.a_player.player.APlayerHeader
import tech.shmy.a_player.player.APlayerInterface
import tech.shmy.a_player.player.APlayerListener
import tech.shmy.a_player.player.APlayerUtil

class AliyunPlayerImpl(context: Context) : APlayerInterface {
    private val aliPlayer: AliPlayer = AliPlayerFactory.createAliPlayer(context)
    private var listener: APlayerListener? = null

    init {
        aliPlayer.setOnVideoSizeChangedListener { width, height ->
            listener?.onVideoSizeChangedListener(width, height)
        }
        aliPlayer.setOnPreparedListener {
            listener?.onInitializedListener()
        }
        aliPlayer.setOnRenderingStartListener {
            listener?.onReadyToPlayListener()
        }
        aliPlayer.setOnStateChangedListener {
            when (it) {
                IPlayer.started -> {
                    listener?.onPlayingListener(true)
                }
                else -> {
                    listener?.onPlayingListener(false)
                }
            }
        }
        aliPlayer.setOnErrorListener {
            listener?.onErrorListener(it.code.name, it.msg)
        }
        aliPlayer.setOnCompletionListener {
            listener?.onCompletionListener()
        }
        aliPlayer.setOnInfoListener { it ->
            when (it.code) {
                InfoCode.CurrentPosition -> {
                    listener?.onCurrentPositionChangedListener(it.extraValue)
                }
                InfoCode.CurrentDownloadSpeed -> {
                    listener?.onCurrentDownloadSpeedChangedListener(it.extraValue)
                }
                InfoCode.BufferedPosition -> {
                    listener?.onBufferedPositionChangedListener(it.extraValue)
                }
                InfoCode.SwitchToSoftwareVideoDecoder -> {
                    listener?.onSwitchToSoftwareVideoDecoderListener()
                }
            }

        }
        aliPlayer.setOnLoadingStatusListener(object : IPlayer.OnLoadingStatusListener {
            override fun onLoadingBegin() {
                listener?.onLoadingBeginListener()
            }

            override fun onLoadingProgress(p0: Int, p1: Float) {
                listener?.onLoadingProgressListener(p0)
            }

            override fun onLoadingEnd() {
                listener?.onLoadingEndListener()
            }
        })
    }

    companion object {
        fun createPlayer(context: Context): APlayerInterface {
            return AliyunPlayerImpl(context)
        }
    }

    override val duration: Long
        get() = aliPlayer.duration
    override val speed: Float
        get() = aliPlayer.speed
    override val isLoop: Boolean
        get() = aliPlayer.isLoop
    override val isAutoPlay: Boolean
        get() = aliPlayer.isAutoPlay

    override fun addListener(listener: APlayerListener): Unit {
        this.listener = listener
    }

    override fun setSurface(surface: Surface) {
        aliPlayer.setSurface(surface)
    }

    override fun play(): Unit {
        aliPlayer.start()
    }

    override fun pause(): Unit {
        aliPlayer.pause()
    }

    override fun stop(): Unit {
        aliPlayer.stop()
    }

    override fun setHttpDataSource(url: String, startAtPositionMs: Long, headers: Array<APlayerHeader>): Unit {
        val config = aliPlayer.config
        config.mMaxBufferDuration = 1000 * 60 * 10
        config.mMaxBackwardBufferDurationMs = 1 * 60 * 10
        var userAgent: String? = null
        var referer: String? = null
        val customHeaders: ArrayList<String> = ArrayList()
        headers.forEach { header ->
            when {
                APlayerUtil.isUserAgentKey(header.key) -> {
                    userAgent = header.value
                }
                APlayerUtil.isRefererKey(header.key) -> {
                    referer = header.value
                }
                else -> {
                    customHeaders.add("${header.key}:${header.value}")
                }
            }
        }
        if (userAgent != null) {
            config.mUserAgent = userAgent
        }
        if (referer != null) {
            config.mReferrer = referer
        }
        config.customHeaders = customHeaders.toTypedArray()
        aliPlayer.config = config
        aliPlayer.volume = 1.0f
        val urlSource = UrlSource()
        urlSource.uri = url
        aliPlayer.setDataSource(urlSource)
        aliPlayer.seekTo(startAtPositionMs)
    }

    override fun setFileDataSource(path: String, startAtPositionMs: Long) {
        TODO("Not yet implemented")
    }

    override fun setAssetDataSource(path: String, startAtPositionMs: Long) {
        TODO("Not yet implemented")
    }

    override fun enableHardwareDecoder(enabled: Boolean) {
        aliPlayer.enableHardwareDecoder(enabled)
    }

    override fun release(): Unit {
        listener = null
        aliPlayer.setSurface(null)
        aliPlayer.stop()
        aliPlayer.release()
    }

    override fun prepare(isAutoPlay: Boolean): Unit {
        aliPlayer.isAutoPlay = isAutoPlay
        aliPlayer.prepare()
    }

    override fun seekTo(positionMs: Long): Unit {
        aliPlayer.seekTo(positionMs, IPlayer.SeekMode.Accurate)
    }

    override fun setSpeed(speed: Float): Unit {
        aliPlayer.speed = speed
    }

    override fun setLoop(isLoop: Boolean): Unit {
        aliPlayer.isLoop = isLoop
    }
}

File: ios/Classes/Player/Impl/AliyunPlayerImpl.swift
//
//  AliyunPlayerImpl.swift
//  a_player
//
//  Created by drm-chao wang on 2022/7/4.
//

import Foundation
import AliyunPlayer

class AliyunPlayerImpl: NSObject, APlayerInterface, AVPDelegate, CicadaRenderDelegate {

   private let aliPlayer: AliPlayer = AliPlayer.init()
   private var listener: APlayerListener?
   var duration: Int64 {
       get {
           return aliPlayer.duration
       }
   }

   var speed: Float {
       get {
           return aliPlayer.rate
       }
   }

   var isLoop: Bool {
       get {
           return aliPlayer.isLoop
       }
   }

   override init() {
       AliPlayerGlobalSettings.setUseHttp2(true)
       super.init()
       aliPlayer.enableHardwareDecoder = true
       aliPlayer.delegate = self
       aliPlayer.renderDelegate = self
   }
   func addListener(listener: APlayerListener) {
       self.listener = listener
   }
   func setHttpDataSource(url: String, startAtPositionMs: Int64, headers: Dictionary<String, String>) {
       let urlSource = AVPUrlSource.init().url(with: url)
       let playerConfig = aliPlayer.getConfig()
       playerConfig?.clearShowWhenStop = true
       playerConfig?.maxBufferDuration = 1000 * 60 * 10
       playerConfig?.mMAXBackwardDuration = 1 * 60 * 10
       var userAgent: String = ""
       var referer: String = ""
       let customHeaders: NSMutableArray = NSMutableArray.init()
       headers.forEach { (key: String, value: String) in
           if (APlayerUtil.isUserAgentKey(key: key)) {
               userAgent = value
           } else if (APlayerUtil.isRefererKey(key: key)) {
               referer = value
           } else {
               customHeaders.add("\(key):\(value)")
           }
       }
       playerConfig?.userAgent = userAgent
       playerConfig?.referer = referer

//       playerConfig?.httpHeaders = customHeaders
       aliPlayer.setConfig(playerConfig)
       aliPlayer.setUrlSource(urlSource)
       aliPlayer.seek(toTime: startAtPositionMs, seekMode: AVP_SEEKMODE_ACCURATE)

   }

   func setFileDataSource(path: String, startAtPositionMs: Int64) {
       setHttpDataSource(url: path, startAtPositionMs: startAtPositionMs, headers: Dictionary<String, String>.init())
   }

   func play() {
       aliPlayer.start()
   }

   func pause() {
       aliPlayer.pause()
   }

   func stop() {
       aliPlayer.stop()
   }

   func destroy() {
       aliPlayer.destroy()
   }

   func prepare() {
       aliPlayer.prepare()
   }

   func seekTo(positionMs: Int64) {
       aliPlayer.seek(toTime: positionMs, seekMode: AVP_SEEKMODE_ACCURATE)
   }

   func setSpeed(speed: Float) {
       aliPlayer.rate = speed
   }

   func setLoop(isLoop: Bool) {
       aliPlayer.isLoop = isLoop
   }

   // CicadaRenderDelegate
   func onVideoPixelBuffer(_ pixelBuffer: CVPixelBuffer!, pts: Int64) -> Bool {
       listener?.onPixelBuffer(pixelBuffer: pixelBuffer)
       return false
   }
   // AVPDelegate
   func onPlayerEvent(_ player: AliPlayer!, eventType: AVPEventType) {
       switch(eventType) {
       case AVPEventPrepareDone:
           listener?.onReadyToPlayListener()
           break
       case AVPEventLoadingStart:
           listener?.onLoadingBeginListener()
           break
       case AVPEventLoadingEnd:
           listener?.onLoadingEndListener()
           break
//       case AVPEventFirstRenderedStart:
//           listener?.onReadyToPlayListener()
//           break
       default:
           break
       }
   }
   func onVideoSizeChanged(_ player: AliPlayer!, width: Int32, height: Int32, rotation: Int32) {
       listener?.onVideoSizeChangedListener(width: width, height: height)
   }
   func onError(_ player: AliPlayer!, errorModel: AVPErrorModel!) {
       listener?.onErrorListener(code: "\(errorModel.code)", message: errorModel.message)
   }

   func onCurrentDownloadSpeed(_ player: AliPlayer!, speed: Int64) {
       listener?.onCurrentDownloadSpeedChangedListener(speed: speed)
   }
   func onCurrentPositionUpdate(_ player: AliPlayer!, position: Int64) {
       listener?.onCurrentPositionChangedListener(position: position)
   }
   func onLoadingProgress(_ player: AliPlayer!, progress: Float) {
       listener?.onLoadingProgressListener(percent: Int(progress))
   }
   func onBufferedPositionUpdate(_ player: AliPlayer!, position: Int64) {
       listener?.onBufferedPositionChangedListener(buffered: position)
   }
   func onPlayerStatusChanged(_ player: AliPlayer!, oldStatus: AVPStatus, newStatus: AVPStatus) {
       if (newStatus == AVPStatusStarted) {
           listener?.onPlayingListener(isPlaying: true)
       } else {
           listener?.onPlayingListener(isPlaying: false)
       }
       if (newStatus == AVPStatusCompletion) {
           listener?.onCompletionListener()
       }
   }
}
